Block production failed: not enough deposits included

### Description

State transition requires that as many deposits as possible are included in each block:
```
    assert len(body.deposits) == min(MAX_DEPOSITS, state.eth1_data.deposit_count - state.eth1_deposit_index)
```

In at least some situations, Teku can fail this condition.  This may be because we hadn't downloaded enough deposits yet or had some other issue with the Eth1 chain.  We should at minimum provide a much clearer error message. If there's anything we can do to reduce the likelihood of missing deposits we should do that too.

```
java.util.concurrent.CompletionException: tech.pegasys.teku.core.StateTransitionException: tech.pegasys.teku.core.exceptions.BlockProcessingException: java.lang.IllegalArgumentException: process_operations: Verify that outstanding deposits are processed up to the maximum number of deposits
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1108) ~[?:?]
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235) ~[?:?]
	at tech.pegasys.teku.util.async.SafeFuture.thenCompose(SafeFuture.java:320) ~[teku-util-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.util.async.SafeFuture.thenApplyChecked(SafeFuture.java:280) ~[teku-util-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.createFromBlockAndState(ValidatorApiHandler.java:154) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.createUnsignedBlock(ValidatorApiHandler.java:141) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at jdk.internal.reflect.GeneratedMethodAccessor80.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at tech.pegasys.teku.events.DirectEventDeliverer.executeMethod(DirectEventDeliverer.java:67) ~[teku-events-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.events.DirectEventDeliverer.deliverToWithResponse(DirectEventDeliverer.java:60) ~[teku-events-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.events.AsyncEventDeliverer.lambda$deliverToWithResponse$1(AsyncEventDeliverer.java:70) ~[teku-events-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.events.AsyncEventDeliverer$QueueReader.deliverNextEvent(AsyncEventDeliverer.java:111) [teku-events-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.events.AsyncEventDeliverer$QueueReader.run(AsyncEventDeliverer.java:103) [teku-events-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: tech.pegasys.teku.core.StateTransitionException: tech.pegasys.teku.core.exceptions.BlockProcessingException: java.lang.IllegalArgumentException: process_operations: Verify that outstanding deposits are processed up to the maximum number of deposits
	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:112) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.BlockProposalUtil.createNewUnsignedBlock(BlockProposalUtil.java:78) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.BlockFactory.createUnsignedBlock(BlockFactory.java:97) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createUnsignedBlock$3(ValidatorApiHandler.java:144) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createFromBlockAndState$4(ValidatorApiHandler.java:159) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.util.async.SafeFuture.lambda$thenApplyChecked$16(SafeFuture.java:283) ~[teku-util-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106) ~[?:?]
	... 16 more
Caused by: tech.pegasys.teku.core.exceptions.BlockProcessingException: java.lang.IllegalArgumentException: process_operations: Verify that outstanding deposits are processed up to the maximum number of deposits
	at tech.pegasys.teku.core.BlockProcessorUtil.process_operations_no_validation(BlockProcessorUtil.java:232) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.StateTransition.lambda$process_block$0(StateTransition.java:130) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.datastructures.state.BeaconStateImpl.updated(BeaconStateImpl.java:273) ~[teku-ethereum-datastructures-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.StateTransition.process_block(StateTransition.java:125) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:97) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.BlockProposalUtil.createNewUnsignedBlock(BlockProposalUtil.java:78) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.BlockFactory.createUnsignedBlock(BlockFactory.java:97) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createUnsignedBlock$3(ValidatorApiHandler.java:144) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createFromBlockAndState$4(ValidatorApiHandler.java:159) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.util.async.SafeFuture.lambda$thenApplyChecked$16(SafeFuture.java:283) ~[teku-util-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106) ~[?:?]
	... 16 more
Caused by: java.lang.IllegalArgumentException: process_operations: Verify that outstanding deposits are processed up to the maximum number of deposits
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142) ~[guava-28.1-jre.jar:?]
	at tech.pegasys.teku.core.BlockProcessorUtil.process_operations_no_validation(BlockProcessorUtil.java:212) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.StateTransition.lambda$process_block$0(StateTransition.java:130) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.datastructures.state.BeaconStateImpl.updated(BeaconStateImpl.java:273) ~[teku-ethereum-datastructures-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.StateTransition.process_block(StateTransition.java:125) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:97) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.core.BlockProposalUtil.createNewUnsignedBlock(BlockProposalUtil.java:78) ~[teku-ethereum-core-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.BlockFactory.createUnsignedBlock(BlockFactory.java:97) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createUnsignedBlock$3(ValidatorApiHandler.java:144) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createFromBlockAndState$4(ValidatorApiHandler.java:159) ~[teku-validator-coordinator-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at tech.pegasys.teku.util.async.SafeFuture.lambda$thenApplyChecked$16(SafeFuture.java:283) ~[teku-util-0.11.3-SNAPSHOT.jar:0.11.3-dev-27fcc726]
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106) ~[?:?]
	... 16 more
```

--- a/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/DepositProvider.java
+++ b/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/DepositProvider.java
@@ -101,6 +101,15 @@ public class DepositProvider implements Eth1EventsChannel, FinalizedCheckpointCh
     UnsignedLong eth1DepositCount = state.getEth1_data().getDeposit_count();
 
     UnsignedLong fromDepositIndex = state.getEth1_deposit_index();
+
+    // We need to have all the deposits that can be included in the state available to ensure
+    // the generated proofs are valid
+    final UnsignedLong lastAvailableDepositIndex =
+        depositNavigableMap.isEmpty() ? fromDepositIndex : depositNavigableMap.lastKey();
+    if (lastAvailableDepositIndex.compareTo(eth1DepositCount) < 0) {
+      throw new MissingDepositsException(lastAvailableDepositIndex, eth1DepositCount);
+    }
+
     UnsignedLong latestDepositIndexWithMaxBlock =
         fromDepositIndex.plus(UnsignedLong.valueOf(MAX_DEPOSITS));
 

--- a/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/MissingDepositsException.java
+++ b/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/MissingDepositsException.java
@@ -1,0 +1,28 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.coordinator;
+
+import com.google.common.primitives.UnsignedLong;
+
+public class MissingDepositsException extends IllegalStateException {
+
+  public MissingDepositsException(
+      final UnsignedLong maxAvailableDeposit, final UnsignedLong requiredDepositIndex) {
+    super(
+        "Unable to create block because ETH1 deposits are not available. Missing deposits "
+            + maxAvailableDeposit.plus(UnsignedLong.ONE)
+            + " to "
+            + requiredDepositIndex);
+  }
+}

--- a/validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/DepositProviderTest.java
+++ b/validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/DepositProviderTest.java
@@ -14,6 +14,7 @@
 package tech.pegasys.teku.validator.coordinator;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
@@ -141,6 +142,29 @@ public class DepositProviderTest {
         .onBlockWithDeposit(
             event.getBlockTimestamp(),
             new Eth1Data(depositMerkleTree.getRoot(), UnsignedLong.ONE, event.getBlockHash()));
+  }
+
+  @Test
+  void shouldThrowMissingDepositsExceptionWhenRequiredDepositsAreNotAvailable() {
+    mockStateEth1DepositIndex(5);
+    mockEth1DataDepositCount(10);
+    assertThatThrownBy(() -> depositProvider.getDeposits(beaconState))
+        .isInstanceOf(MissingDepositsException.class)
+        .hasMessageContaining("6 to 10");
+  }
+
+  @Test
+  void shouldThrowMissingDepositsExceptionWhenAllDepositsRequiredForStateNotAvailable() {
+    // To generate a valid proof we need the deposits up to state deposit count
+    // So fail even if we could have filled MAX_DEPOSITS
+    Constants.MAX_DEPOSITS = 1;
+    mockDepositsFromEth1Block(0, 8);
+    mockStateEth1DepositIndex(5);
+    mockEth1DataDepositCount(10);
+
+    assertThatThrownBy(() -> depositProvider.getDeposits(beaconState))
+        .isInstanceOf(MissingDepositsException.class)
+        .hasMessageContaining("8 to 10");
   }
 
   private void checkThatDepositProofIsValid(SSZList<Deposit> deposits) {

